python/cuml/common/import_utils.py

```diff
@@ -126,18 +127,33 @@ def check_cupy8(conf=None):
     Will otherwise raise an error in case CuPy 8.0+ is unavailable.
 
     """
-    def check_cupy8_dec(func):
-        @wraps(func)
+
+    def get_inner(f):
+
+        @wraps(f)
         def inner(*args, **kwargs):
             import cupy as cp
             if LooseVersion(str(cp.__version__)) >= LooseVersion('8.0'):
-                return func(*args, **kwargs)
+                return f(*args, **kwargs)
             else:
                 err_msg = 'Could not import required module CuPy 8.0+'
                 if conf == 'pytest':
                     import pytest
                     pytest.skip(err_msg)
                 else:
                     raise ImportError(err_msg)
+
         return inner
+
+    def check_cupy8_dec(func):
+
+        # If this is a class, we dont want to wrap the class which will turn it
+        # into a function. Instead, wrap __new__ to have the same effect
+        if (inspect.isclass(func)):
+
+            func.__new__ = get_inner(func.__new__)
```

Comment:
should we just not wrap classes with this decorator?

Reply:
That depends on how we want want the module import to work when the user doesn't have the required features. Right now everything will import fine but will only throw an error if you try to use (i.e. create an instance of the class or call a function) a class or method decorated with `check_cupy8`

If we like that functionality, I think we need to decorate the classes with this decorator.

Reply:
I think we want the import to work when cupy8 is missing but only to trigger an error when they call a relevant function. There are some preprocessing modules that work on cupy 7 as well. I thought that was how it worked previously?

Reply:
Ok, I see the misunderstanding. I didnt change the functionality at all from before. It still works like it used to, I just updated it to handle classes better. When using a decorator on a class, its better to wrap the individual method of the class instead of wrapping the class itself. Wrapping the class itself turns it into a function that returns a class. For example, in branch 0.16:

```python
>>> type(cuml.experimental.preprocessing.Normalizer)
<class 'function'>
>>> inspect.isclass(cuml.experimental.preprocessing.Normalizer)
False
```
And in this PR:
```python
>>> type(cuml.experimental.preprocessing.Normalizer)
<class 'type'>
>>> inspect.isclass(cuml.experimental.preprocessing.Normalizer)
True
```

Sphinx needs to know the type for autodoc to function properly.

Reply:
Reverted back close to its original state. @JohnZed I added a single assert line to ensure the decorator isnt used on any classes that should prevent this issue in the future. Let me know what you think. https://github.com/rapidsai/cuml/pull/2942/files#diff-52436e61f2011deef045c8e06ffb2003R132

Reply:
I like it!